### PR TITLE
refactor(range): implement component

### PR DESCRIPTION
**Pull request by aesteves60**

### CODE REVIEW

**Review by dpellier — CHANGES_REQUESTED** [review]
Demo:
- if I change the `max` to a lower value than the current one, the tooltip does not update to the max value but keep the bigger one
- issue on the custom CSS example (tested on FF):
Capture d’écran 2024-07-02 à 09 31 02

- playing with step on dual leads to weird state:
Capture d’écran 2024-07-02 à 09 44 52

**Review by astagnol — CHANGES_REQUESTED** [review]
### Global
- we should not be able to click on the trackbar anymore (for both single and dual)
- "Dual: do not allow each handle to cross. They can be put on the same value, then max always stays on top"

**Review by feoche — CHANGES_REQUESTED** [review]
## Global
- [ ] Default value should be : `min:0`, `max:100` (as native Range does)
- [x] When no `value` is defined, it should have the same strategy as native Range does - see https://developer.mozilla.org/fr/docs/Web/HTML/Element/input/range#valeur
- [ ] There are misalignments between `ods-range`'s `value` and the inner's `input`'s `value` (especially tweaking `min` and `max` bounds)
- [ ] Range trackbar should not be clickable
- [ ] Dual Range : Values should keep a logical order, meaning we can't have a max bound before a min bound in DOM (issue especially on focus)
- [ ] Dual Range : setting `min` and/or `max` inside value bounds (thumbs) hides one of them

## Style
- [ ] Setting `isDisabled` and `hasError` to `true` should not display the error style (only disabled)
- [ ] Changing `step` messes up the thumb position on the Range
- [ ] Setting `min` to something greater than `value` breaks
- [ ] Setting `max` to something less than `value` puts Tooltip outside of Range

## Documentation
- [ ] All examples : Add extra top margin for tooltip
- [ ] Style customization : Messes up the range 
image

File: packages/storybook/stories/components/range/documentation.mdx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+<Canvas of={ RangeStories.Error } sourceState='shown' />
+
+## Max
+<Canvas of={ RangeStories.Max } sourceState='shown' />
+
+## Min
+<Canvas of={ RangeStories.Min } sourceState='shown' />
+
+## Step
+<Canvas of={ RangeStories.Step } sourceState='shown' />
```
Comment (dpellier) [inline packages/storybook/stories/components/range/documentation.mdx:49]: would be nice to add a dual example too

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by dpellier — CHANGES_REQUESTED** [review]
Capture d’écran 2024-07-15 à 15 16 21

Doc generator seems unable to manage `[number, number]`, either try to fix it (if not much trouble) or use the `/** @docType ` tag

File: packages/ods/src/components/range/src/components/ods-range/ods-range.tsx [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  @Watch('value')
+  private onValueChange(): void {
+    this.isDualRange = isDualRange(this.value);
+    isDualRange(this.value) ? this.changeValues(this.value[0], this.value[1]) : this.changeValues(this.value ?? undefined);
+    setFormValue(this.internals, this.value);
+  }
+
+  onInputElChange(mutationList: MutationRecord[]): void {
```
Comment (dpellier) [inline packages/ods/src/components/range/src/components/ods-range/ods-range.tsx:109]: put after lifecycle and set `private`

File: packages/ods/src/components/range/tests/behavior/ods-range.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
+import type { E2EElement, E2EPage } from '@stencil/core/testing';
+import { newE2EPage } from '@stencil/core/testing';
+
+describe('ods-range behavior', () => {
```
Comment (dpellier) [inline packages/ods/src/components/range/tests/behavior/ods-range.e2e.ts:4]: we went with the `behaviour` style for now, please update the directory name accordingly

File: packages/ods/src/style/_range.scss [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        var(--ods-color-neutral-600) min(var(--ods-range-value-percentage), var(--ods-range-value-percentage-dual)),
+        var(--ods-color-neutral-600) max(var(--ods-range-value-percentage), var(--ods-range-value-percentage-dual)),
+        $range-background-color max(var(--ods-range-value-percentage), var(--ods-range-value-percentage-dual)),
+        $range-background-color 100%);
+    }
+  }
+
+  .ods-range__range {
+    position: absolute;
+    z-index: 2;
```
Comment (dpellier) [inline packages/ods/src/style/_range.scss:151]: if this value is actually meant to be higher than the `z-index: 1` of `.ods-range__range-dual`, use a variable on the latest and set this to `$the-var + 1`

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/ods/src/components/range/src/components/ods-range/ods-range.tsx [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  @Method()
+  async getValidity(): Promise<ValidityState | undefined> {
+    return this.inputEl?.validity;
+  }
+
+  @Method()
+  async reset(): Promise<void> {
+    this.value = this.defaultValue ?? this.min;
```
[the remark this answers is not in the document]
Comment (aesteves60) [reply packages/ods/src/components/range/src/components/ods-range/ods-range.tsx:63]: There a a weird when I set the value to null, the thumb is in the middel but there are no color on the track
image

File: packages/ods/src/components/range/src/components/ods-range/ods-range.tsx [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      if (this.min > valueMin) {
+        this.value = [this.min, this.value[1]];
+      }
+    }
+  }
+
+  @Watch('value')
+  private onValueChange(): void {
+    this.isDualRange = isDualRange(this.value);
+    isDualRange(this.value) ? this.changeValues(this.value[0], this.value[1]) : this.changeValues(this.value ?? undefined);
```
[the remark this answers is not in the document]
Comment (aesteves60) [reply packages/ods/src/components/range/src/components/ods-range/ods-range.tsx:105]: Yeah but if I use `this.isDualRange` typescript don't use the guard and don't know the type of `value`

File: packages/ods/src/components/range/src/components/ods-range/ods-range.tsx [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    if (!isDualRange(this.value)) {
+      if (this.min > this.value) {
+        this.value = this.min;
+      }
+
+      if (this.value > this.max) {
+        this.value = this.max;
+      }
```
[the remark this answers is not in the document]
Comment (aesteves60) [reply packages/ods/src/components/range/src/components/ods-range/ods-range.tsx]: Don't know this very usefull 👍

**Review by astagnol — APPROVED** [review]

File: packages/ods/src/components/range/src/components/ods-range/ods-range.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (dpellier) [reply packages/ods/src/components/range/src/components/ods-range/ods-range.tsx]: Ok as it is the default browser rendering on empty value, but hide the tooltip if value is null

**Review by dpellier — APPROVED** [review]

**Review by feoche — APPROVED** [review]

**Pull request merged**